integration-libs/opf/cta/core/services/opf-dynamic-cta.service.ts

```diff
@@ -0,0 +1,251 @@
+
+  fillCtaRequestforCartPage(
+    scriptLocation: CtaScriptsLocation,
+    paymentAccountIds: number[]
+  ) {
+    this.isCartPage = true;
+    return combineLatest({
+      cart: this.activeCartFacade.takeActive(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ cart, additionalData }) => {
+        return {
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          ctaProductItems: cart.entries?.map((entry) => {
+            return {
+              productId: entry.product?.code,
+              quantity: entry.quantity,
+            } as CtaProductItem;
+          }),
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  fillCtaRequestforProductPage(
+    scriptLocation: CtaScriptsLocation,
+    paymentAccountIds: number[]
+  ) {
+    this.isCartPage = false;
+    return combineLatest({
+      product: this.currentProductService.getProduct(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ product, additionalData }) => {
+        return {
+          ctaProductItems: [{ productId: product?.code, quantity: 1 }],
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  protected fillAdditionalData(): Observable<Array<KeyValuePair>> {
+    return combineLatest({
+      language: this.languageService.getActive(),
+      currency: this.currencyService.getActive(),
+    }).pipe(
+      take(1),
+      map(({ language, currency }) => {
+        return [
+          {
+            key: 'locale',
+            value: language,
+          },
+          {
+            key: 'currency',
+            value: currency,
+          },
+          {
+            key: 'scriptIdentifier',
+            value: this.getNewScriptIdentifier(),
+          },
+        ];
+      })
+    );
+  }
+
+  initiateEvents() {
+    if (!this.isOnsiteMessagingInit) {
+      this.listenScriptReadyEvent();
+
+      this.isCartPage && this.cartChangedListener();
+      !this.isCartPage && this.productChangedListener();
+
+      this.isOnsiteMessagingInit = true;
+    }
+  }
+
+  stopEvents() {
+    if (this.isOnsiteMessagingInit) {
+      this.subList.forEach((sub) => sub.unsubscribe());
+      this.subList = [];
+      this.globalFunctionsFacade.removeGlobalFunctions(
+        GlobalFunctionsDomain.GLOBAL
+      );
+      this.isOnsiteMessagingInit = false;
+    }
+  }
+
+  protected getNewScriptIdentifier(): string {
+    this.scriptIdentifiers.push(String(this.scriptIdentifiers.length + 1));
+    return String(
+      this.scriptIdentifiers[this.scriptIdentifiers.length - 1]
+    ).padStart(4, '0');
+  }
+
+  registerScriptReadyEvent() {
+    this.globalFunctionsFacade.registerGlobalFunctions({
+      paymentSessionId: '',
+      domain: GlobalFunctionsDomain.GLOBAL,
+    });
+  }
```

Comment:
[MEDIUM]
1. Does the registration of the Global Function really have to be called before `CtaScriptsService.fetchCtaScripts()`? If not, I'm in favor of removing this public method, and move the responsibility of registering Global Function to inside of `initiateEvents()`  method - to be analogical to `stopEvents()` method, which already takes care of removing the global function. In other words: if not necessary, I'd like to not leak the responsibility of registering the global function to outside of this class
2. If we _really_ have to expose a public method to be called from outside (which I'm hoping maybe we don't have to), then I'd sugest at least renaming this method  (`registerScriptReadyEvent()` ->`registerGlobalFunction()`). Why: as far as I can understand, this method in its implementation doesn't not touch the scriptReady event - it neither subscribes to it, nor dispatches it. But it regsters a global function.